src/util/bip32.rs

```diff
@@ -210,12 +210,31 @@ impl serde::Serialize for ChildNumber {
     }
 }
 
+/// Trait that allows possibly failable conversion from a type into a
+/// derivation path
+pub trait IntoDerivationPath {
+    /// Convers a given type into a [`DerivationPath`] with possible error
+    fn into_derivation_path(self) -> Result<DerivationPath, Error>;
+}
+
 /// A BIP-32 derivation path.
 #[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
 pub struct DerivationPath(Vec<ChildNumber>);
 impl_index_newtype!(DerivationPath, ChildNumber);
 serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
 
+impl Default for DerivationPath {
+    fn default() -> DerivationPath {
+        DerivationPath::master()
+    }
+}
+
+impl<T> IntoDerivationPath for T where T: Into<DerivationPath> {
```

Comment:
Isn't this one of those constructions where whenever you implement `IntoDerivationPath` for a type, cargo will complain "can't allow this impl because rust-bitcoin can implement `From<T>` for `DerivationPath` and then this implementation is conflicting"? I bumped into that before.

@dr-orlovsky did you try this?

Reply:
There is a good solution for that which I use widely: trait implementation strategies and holder. Proposed by @Kixunil. 

A sample can be found here:
https://github.com/LNP-BP/rust-lnpbp/blob/master/src/paradigms/strict_encoding.rs#L298-L445

Reply:
And, actually, with this downstreams would not need to impl `IntoDerivationPath` at all, just do a `Into<DerivationPath` impl and everything else will work

Reply:
do you have a link to this trait thing?

Reply:
> But did you try to implement `IntoDerivationPath` on a foreign type already?

No, because its not recommended. One should implement `Into<DerivationPath>` instead, like with `IntoString`: you never do it yourself but get that for free from `Display`.

Reply:
> do you have a link to this trait thing?

Sorry, which one? Strategy? I gave it above

Reply:
There are three problems we are mixing here.

1. Implementing foreign traits for foreign types. There is nothing specific for this problem with this commit, it's a common rust issue and the only way of addressing it is to do either wrapped data type or do a new marker trait with generic dependency on the foreign one (like `pub trait NewOne where Self: ForeignOne { /* repeat all methods from ForeignOne */ }`) and implement it on foreign data types. I do not know why and how we started discussion this problem, b/c again there is no specific in this commit which makes it anyhow harder to solve.

2. Implementing trait for a generic type ("blanket implementation") more than once (applies both for local and foreign traits) - or implement foreign trait for a concrete type where there was already some blanked implementation done. The solution is to use special pattern by @Kixunil. I use it widely and have a special helper type here: https://github.com/LNP-BP/rust-amplify/blob/master/src/strategy.rs#L26-L45

With that helper type you can write the following code, which will provide you with efficiently multiple blanket implementations of some trait `SampleTrait`:
```rust
    // Define strategies, one per specific implementation that you need, either blanket or concrete
    pub struct StrategyA;
    pub struct StrategyB;
    pub struct StrategyC;

    // Define a single marker type
    pub trait Strategy {
        type Strategy;
    }

    // Do a single blanket implementation using Holder and Strategy marker trait
    impl<T> SampleTrait for T
    where
        T: Strategy + Clone,
        amplify::Holder<T, <T as Strategy>::Strategy>: SampleTrait,
    {
         // Do this for each of sample trait methods:
         fn sample_trait_method(&self) {
             amplify::Holder::new(self.clone()).sample_trait_method()
         }
    }

    // Do this type of implementation for each of the strategies
    impl<T> SampleTrait for amplify::Holder<T, StrategyA>
    where
        T: StrategyA
    { /* ... write your implementation-specific code here */ }
```

3. Implementing type `IntoDerivationPath` from this PR on some concrete type in downstream dependency. This should not be done, instead one should implement `Into<DerivationPath>` and will get `into_derivation_path()` method for free thanks for a blanket implementation from this PR, and, if for any wired (and unrecommended) reason he'd like to do a direct implementation of `IntoDerivationPath` to some concrete type or another blanket, he should use solution from pt.2 in that case.

Reply:
Heh, sounds like I should move my ass and finally write that article about the pattern which I planned for some time. :)

Reply:
I think the main problem with downstream only being able to implement `Into<DerivationPath>` is that it doesn't allow failure. Idk if that's a blocker though. Imo this trait is a fixme anyway and should be deprecated as soon as MSRV >= 1.34.0, at which point this wouldn't be a problem anymore.